src/proxy-worker.ts

```diff
@@ -64,28 +64,38 @@ class Tunnel extends nodeTunnel.Tunnel {
 		req: nodeTunnel.Request,
 	) => {
 		const { uuid, auth } = this.parseRequest(req);
-		try {
-			await dns.lookup(`${uuid}.vpn`);
-			this.logger.info(`connecting to ${host}:${port}`);
-			const socket = await super.connect(port, host, client, req);
-			metrics.inc(Metrics.ActiveTunnels);
-			metrics.inc(Metrics.TotalTunnels);
-			socket.on('close', () => {
-				metrics.dec(Metrics.ActiveTunnels);
-			});
-			return socket;
-		} catch {
-			// The lookup failed so we try to forward to the correct vpn instance instead
+
+		const deviceIsLocal = async () => {
 			try {
+				await dns.lookup(`${uuid}.vpn`);
+				return true;
+			} catch {
+				return false;
+			}
+		};
+
+		try {
+			if (await deviceIsLocal()) {
+				this.logger.info(`connecting to ${host}:${port}`);
+				const socket = await super.connect(port, host, client, req);
+				metrics.inc(Metrics.ActiveTunnels);
+				metrics.inc(Metrics.TotalTunnels);
+				socket.on('close', () => {
+					metrics.dec(Metrics.ActiveTunnels);
+				});
+				return socket;
+			} else {
 				const vpnHost = await device.getDeviceVpnHost(uuid, auth);
 				if (vpnHost.id === this.serviceId) {
+					client.end('HTTP/1.0 500 Internal Server Error\r\n\r\n');
```

Comment:
It might make sense to break this out to a const since it's repeated a bunch of times